core/src/main/java/org/kiwix/kiwixmobile/core/history/viewmodel/HistoryViewModel.kt

```diff
@@ -0,0 +1,217 @@
+  val actions = PublishProcessor.create<Action>()
+  private val filter = BehaviorProcessor.createDefault("")
+  private val compositeDisposable = CompositeDisposable()
+  private val currentBook = BehaviorProcessor.createDefault("")
+  private val showAllSwitchToggle = BehaviorProcessor.createDefault(false)
+  private val deselectAllItems = BehaviorProcessor.createDefault(false)
+  private var isInSelectionMode = false
+
+  init {
+    compositeDisposable.addAll(viewStateReducer(), actionMapper(), updateSelectionMode())
+  }
+
+  private fun updateSelectionMode() =
+    deselectAllItems.subscribe({ isInSelectionMode = it }, Throwable::printStackTrace)
+
+  private fun viewStateReducer() = Flowable.combineLatest(
+    currentBook,
+    searchResults(),
+    deselectAllItems,
+    filter,
+    showAllSwitchToggle,
+    Function5(::updateResultsState)
+  ).subscribe(state::postValue, Throwable::printStackTrace)
+
+  private fun searchResults(): Flowable<List<HistoryListItem>> =
+    Flowable.combineLatest(
+      filter,
+      showAllSwitchToggle,
+      historyDao.history(),
+      Function3(::searchResults)
+    )
+
+  private fun searchResults(
+    searchString: String,
+    showAllToggle: Boolean,
+    historyList: List<HistoryListItem>
+  ): List<HistoryListItem> = historyList
+      .filterIsInstance<HistoryItem>()
+      .filter { h ->
+        h.historyTitle.contains(searchString, true) &&
+          (h.zimName == zimReaderContainer.name || showAllToggle)
+      }
+
+  private fun updateResultsState(
+    currentBook: String,
+    historyItemSearchResults: List<HistoryListItem>,
+    shouldDeselectAllItems: Boolean,
+    searchString: String,
+    showAllSwitchOn: Boolean
+  ): State {
+    if (shouldDeselectAllItems) {
+      deselectAllHistoryItems(historyItemSearchResults)
+    }
+    val selectedItems = filterOutAllSelectedHistoryItems(historyItemSearchResults)
+    val historyListWithDateItems = addDateHeadersToHistoryItems(historyItemSearchResults)
+    return selectStateBasedOnInput(
+      currentBook,
+      historyListWithDateItems,
+      selectedItems,
+      searchString,
+      showAllSwitchOn
+    )
+  }
+
+  private fun selectStateBasedOnInput(
+    currentBook: String,
+    historyListWithDateItems: List<HistoryListItem>,
+    selectedItems: List<HistoryListItem>,
+    searchString: String,
+    showAllSwitchOn: Boolean
+  ) = when {
+    historyListWithDateItems.isEmpty() -> NoResults(searchString, historyListWithDateItems)
+    selectedItems.isNotEmpty() -> SelectionResults(
+      searchString,
+      historyListWithDateItems,
+      selectedItems,
+      showAllSwitchOn,
+      currentBook
+    )
+    else -> Results(searchString, historyListWithDateItems, showAllSwitchOn, currentBook)
+  }
+
+  private fun addDateHeadersToHistoryItems(historyItemSearchResults: List<HistoryListItem>):
+    MutableList<HistoryListItem> {
+    return historyItemSearchResults.reversed().foldOverAddingHeaders(
```

Comment:
should these be reversed? I haven't actually run this on a device just yet but if most recent is not being returned first we can probably change the call in the DAO to give us what we want

Reply:
For some reason I had to reverse them for them to be in order, but only when using the new foldOverAddingHeaders. We did not have to reverse them when the foldOverAddingHeaders was used in the previous historyActivity.java.

Reply:
I changed this so that the items are sorted in order to make testing easier. I think it looks nicer as well. It also guarantees an ordered list if historyDao has the items out of order (very unlikely?).